test/test_python_dispatch.py

```diff
@@ -497,5 +497,187 @@ def wrap(e):
+
+    elem: torch.Tensor
+
+    __slots__ = ['elem', 'elem_']
+
+    @staticmethod
+    def __new__(cls, elem, *args, **kwargs):
+        # r = torch.Tensor._make_wrapper_subclass(cls, elem)
+        r = torch.Tensor._make_wrapper_subclass(
+            cls, elem.size(), elem.stride(),
+            dtype=elem.dtype, layout=elem.layout,
+            device='meta', requires_grad=elem.requires_grad
+        )
+        if elem._is_functional_tensor():
+            r.elem_ = elem
+            r.elem = r.elem_
+        else:
+            r.elem_ = elem
+            r.elem = r.elem_._to_functional_tensor()
+        return r
+
+    def __repr__(self):
+        assert self.elem._is_functional_tensor()
+        self.elem._sync()
+        return f"FunctionalTensor({self.elem._from_functional_tensor()})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e):
+            if not isinstance(e, torch.Tensor):
+                return e
+            if isinstance(e, FunctionalTensor):
+                return e.elem
+            # This can happen if I do FunctionalTensor(torch.ones(2)) + torch.ones(2)
+            assert not e._is_functional_tensor()
+            return e._to_functional_tensor()
+
+        def wrap(e):
+            if not isinstance(e, torch.Tensor):
+                return e
+            assert not isinstance(e, FunctionalTensor)
+            assert e._is_functional_tensor()
+            return FunctionalTensor(e)
+
+        unwrapped_args = tree_map(unwrap, args)
+        if func.__name__ == 'add_':
+            unwrapped_res = unwrapped_args[0].add_(*unwrapped_args[1:], **kwargs)
+        else:
+            unwrapped_res = func(*unwrapped_args, **kwargs)
+        res = tree_map(wrap, unwrapped_res)
+        return res
+
+class TestFunctionalization(TestCase):
+
+    def test_functionalization(self):
+        def f5(x, *, logging: bool):
+            # test: everything
+            with capture_logs() as logs:
+                tmp = torch.ones(2, 2)
+                y = x.view(8)
+                z0 = y.reshape(2, 4)
+                z1 = z0.transpose(1, 0)
+                z1.unsqueeze_(0)
+                z1.squeeze_()
+                z2, z3 = z1.split(2)
+                z2.add_(tmp)
+            if logging:
+                self.assertExpectedInline('\n'.join(logs), """\
+$1 = torch._ops.aten.view($0, [8])
+$2 = torch._ops.aten._reshape_alias($1, [2, 4], [4, 1])
+$3 = torch._ops.aten.transpose($2, 1, 0)
+$4 = torch._ops.aten.view($0, [8])
+$5 = torch._ops.aten._reshape_alias($4, [2, 4], [4, 1])
+$6 = torch._ops.aten.transpose($5, 1, 0)
+$7 = torch._ops.aten.unsqueeze($6, 0)
+$8 = torch._ops.aten.view($0, [8])
+$9 = torch._ops.aten._reshape_alias($8, [2, 4], [4, 1])
+$10 = torch._ops.aten.transpose($9, 1, 0)
+$11 = torch._ops.aten.unsqueeze($10, 0)
+$12 = torch._ops.aten.squeeze($11)
+$13, $14 = torch._ops.aten.split($12, 2)
+$15 = torch._ops.aten.add($13, tensor([[1., 1.],
```

Comment:
How come there aren't a bunch of adds for all the other intermediate quantities here?